### PR TITLE
Issue #7608: Update doc for DesignForExtension

### PR DESCRIPTION
**Pull request by zwwang98**
# Issue #7608: Update doc for DesignForExtension

[x] Updated doc starts with default config and example, followed by properties and their examples
[x] Passed that command `mvn clean verify` pass without violations
[x] Screenshots of how website looks like are attached below
[x] CLI results of different examples are attached below

## Screenshots of website
<img width="616" alt="Screen Shot 2023-03-27 at 7 51 17 PM" src="https://user-images.githubusercontent.com/29494959/228149621-5dacdce9-66fb-4d39-a676-40602aeb5541.png">
<img width="630" alt="Screen Shot 2023-03-27 at 7 52 00 PM" src="https://user-images.githubusercontent.com/29494959/228149697-be09c0d8-ab6a-4b58-8c6b-e34aaf2c8be6.png">
<img width="691" alt="Screen Shot 2023-03-27 at 7 52 15 PM" src="https://user-images.githubusercontent.com/29494959/228149732-9ea905e2-8262-4c0d-9d97-7d706b565ee4.png">
<img width="841" alt="Screen Shot 2023-03-27 at 7 52 33 PM" src="https://user-images.githubusercontent.com/29494959/228149775-96fbbdc3-5bfe-4edd-a2b5-afbe997d43b4.png">

## CLI Results for Examples
### Example for Default Config
```terminal
(base) a9527@9527deMacBook-Pro java % javac Foo.java
(base) a9527@9527deMacBook-Pro java % cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="DesignForExtension"/>
  </module>
</module>
(base) a9527@9527deMacBook-Pro java % cat Foo.java  
public abstract class Foo {
  private int bar;

  public int method1() {return 2;}

  public int method2() {return 8;}

  private void method3() {method4();}

  protected void method4() { }

  public abstract void method5();
  /**
   * This implementation ...
   @return some int value.
   */
  public int method6() {return 1;}

  /**
   * Some comments ...
   */
  public int method7() {return 1;}

  /**
   * This
   * implementation ...
   */
  public int method8() {return 2;}

  @Override
  public String toString() {
    return "Foo{" +
        "bar=" + bar +
        '}';
  }
}
(base) a9527@9527deMacBook-Pro java % java -classpath ../../../out/artifacts/learn_checkstyle_jar/learn-checkstyle.jar:../../../../libcheckstyle-10.9.2-all.jar com.puppycrawl.tools.checkstyle.Main -c ./config.xml Foo.java 
Starting audit...
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:4:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'method1' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'method1' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:6:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'method2' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'method2' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:30:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'toString' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'toString' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
Audit done.
Checkstyle ends with 3 errors.      
```

### Example for Config Property `igonredAnnotations`
```
(base) a9527@9527deMacBook-Pro java % javac Foo.java                                                                                                                                                                          
(base) a9527@9527deMacBook-Pro java % cat config.xml                                                                                                                                                                          
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="DesignForExtension">
      <property name="ignoredAnnotations" value="Override"/>
    </module>
  </module>
</module>
(base) a9527@9527deMacBook-Pro java % cat Foo.java                                                                                                                                                                            
public abstract class Foo {
  private int bar;

  public int method1() {return 2;}

  public int method2() {return 8;}

  private void method3() {method4();}

  protected void method4() { }

  public abstract void method5();
  /**
   * This implementation ...
   @return some int value.
   */
  public int method6() {return 1;}

  /**
   * Some comments ...
   */
  public int method7() {return 1;}

  /**
   * This
   * implementation ...
   */
  public int method8() {return 2;}

  @Override
  public String toString() {
    return "Foo{" +
        "bar=" + bar +
        '}';
  }
}%                                                                                                                                                                                                                                      (base) a9527@9527deMacBook-Pro java % java -classpath ../../../out/artifacts/learn_checkstyle_jar/learn-checkstyle.jar:../../../../libcheckstyle-10.9.2-all.jar com.puppycrawl.tools.checkstyle.Main -c ./config.xml Foo.java 
Starting audit...
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:4:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'method1' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'method1' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:6:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'method2' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'method2' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
Audit done.
Checkstyle ends with 2 errors.
```

### Example for Config Property `requiredJavadocPhrase` (one line)
```
(base) a9527@9527deMacBook-Pro java % javac Foo.java                                                                                                                                                                          
(base) a9527@9527deMacBook-Pro java % cat config.xml                                                                                                                                                                          
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="DesignForExtension">
      <property name="requiredJavadocPhrase" value="This implementation"/>
    </module>
  </module>
</module>
(base) a9527@9527deMacBook-Pro java % cat Foo.java                                                                                                                                                                            
public abstract class Foo {
  private int bar;

  public int method1() {return 2;}

  public int method2() {return 8;}

  private void method3() {method4();}

  protected void method4() { }

  public abstract void method5();
  /**
   * This implementation ...
   @return some int value.
   */
  public int method6() {return 1;}

  /**
   * Some comments ...
   */
  public int method7() {return 1;}

  /**
   * This
   * implementation ...
   */
  public int method8() {return 2;}

  @Override
  public String toString() {
    return "Foo{" +
        "bar=" + bar +
        '}';
  }
}%                                                                                                                                                                                                                                      (base) a9527@9527deMacBook-Pro java % java -classpath ../../../out/artifacts/learn_checkstyle_jar/learn-checkstyle.jar:../../../../libcheckstyle-10.9.2-all.jar com.puppycrawl.tools.checkstyle.Main -c ./config.xml Foo.java 
Starting audit...
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:4:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'method1' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'method1' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:6:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'method2' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'method2' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:22:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'method7' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'method7' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:28:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'method8' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'method8' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:30:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'toString' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'toString' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
Audit done.
Checkstyle ends with 5 errors.
```

### Example for Config Property `requiredJavadocPhrase` (span in multiple line)
```
(base) a9527@9527deMacBook-Pro java % javac Foo.java                                                                                                                                                                          
(base) a9527@9527deMacBook-Pro java % cat config.xml                                                                                                                                                                          
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="DesignForExtension">
      <property name="requiredJavadocPhrase" value="This[\s\S]*implementation"/>
    </module>
  </module>
</module>
(base) a9527@9527deMacBook-Pro java % cat Foo.java                                                                                                                                                                            
public abstract class Foo {
  private int bar;

  public int method1() {return 2;}

  public int method2() {return 8;}

  private void method3() {method4();}

  protected void method4() { }

  public abstract void method5();
  /**
   * This implementation ...
   @return some int value.
   */
  public int method6() {return 1;}

  /**
   * Some comments ...
   */
  public int method7() {return 1;}

  /**
   * This
   * implementation ...
   */
  public int method8() {return 2;}

  @Override
  public String toString() {
    return "Foo{" +
        "bar=" + bar +
        '}';
  }
}%                                                                                                                                                                                                                                      (base) a9527@9527deMacBook-Pro java % java -classpath ../../../out/artifacts/learn_checkstyle_jar/learn-checkstyle.jar:../../../../libcheckstyle-10.9.2-all.jar com.puppycrawl.tools.checkstyle.Main -c ./config.xml Foo.java 
Starting audit...
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:4:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'method1' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'method1' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:6:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'method2' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'method2' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:22:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'method7' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'method7' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /Users/CS/learn-checkstyle/src/main/java/Foo.java:30:3: Class 'Foo' looks like designed for extension (can be subclassed), but the method 'toString' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'Foo' final or making the method 'toString' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
Audit done.
Checkstyle ends with 4 errors.
```

### CODE REVIEW

**Comment by zwwang98** [comment]
> GitHub, generate site

Hi, I don't really understand your reply. Do you mean I should generate website locally and then it will pass the only unsuccessful check? Thanks in advance!

**Comment by romani** [comment]
It is command to GitHub to do a work

**Comment by zwwang98** [comment]
> It is command to GitHub to do a work

Oh I see, I still got a lot to learn :smile:.

Also, could you help me figure out why all checks are passed except the one about checkstyle.checkstyle (Job MacOS JDK17 verify)? I checked Azure pipelines error message, all they have is 

```
##[error]The job running on agent Azure Pipelines 3 ran longer than the maximum time of 60 minutes. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134
,##[warning]Agent Azure Pipelines 3 did not respond to a cancelation request with 00:05:00.
```

Does this have anything to do with my code, or it is just about the Azure pipeline's health status?

Thank you for your time and patience!

**Comment by romani** [comment]
We will rerun it if it not related to your code change

**Comment by zwwang98** [comment]
> We will rerun it if it not related to your code change

Thank you so much for your prompt reply. I really appreciate it, and I didn't expect you to respond so quickly. Thank checkstyle for providing such a valuable opportunity! 😄

**Comment by romani** [comment]
GitHub, generate site

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java [hunk 1]
```diff
@@ -175,45 +175,93 @@
  * <pre>
  * &lt;module name=&quot;DesignForExtension&quot;/&gt;
  * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * public abstract class Foo {
+ *   private int bar;
+ *
+ *   public int method1() {return 2;}  // Violation. No javadoc.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java:183]: Make all method name to be like m1, m2,...

We need code more horizontally compact to see violation comment on phone screen.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java [hunk 2]
```diff
@@ -175,45 +175,93 @@
...
+ *
+ *   &#47;**
+ *    * This
+ *    * implementation ...
+ *    *&#47;
+ *   public int method8() {return 2;}  // OK. Have javadoc on overridable method
+ *
+ *   &#64;Override
+ *   public String toString() {  // Violation. No javadoc for @Override method
+ *     return "Foo{" +
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java:212]: `return "";`

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java (hunk 1 above)
Comment (zwwang98) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java:183]: Thank you! I will do that as soon as I get home!

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java (hunk 2 above)
Comment (zwwang98) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java:212]: Thank you! I will do that as soon as I get home!

**Comment by zwwang98** [comment]
Hi @romani , I already resolve this one in commit [184f6da](https://github.com/zwwang98/checkstyle/commit/184f6da3f50b2d5e26cc0765fd165a23ae29f7e8). Should I create a new Pull Request. Thank you again for being so supportive and for everything else!

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java (hunk 1 above)
Comment (zwwang98) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java]: Hi @romani , I already resolve this one in commit [184f6da](https://github.com/zwwang98/checkstyle/commit/184f6da3f50b2d5e26cc0765fd165a23ae29f7e8).

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java (hunk 2 above)
Comment (zwwang98) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java]: Hi @romani , I already resolve this one in commit [184f6da](https://github.com/zwwang98/checkstyle/commit/184f6da3f50b2d5e26cc0765fd165a23ae29f7e8).

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java (hunk 1 above)
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java]: @zwwang98 allow reviewer to resolve all items as they come back to PR, simply reply "done" to each item as you complete it.

Comment (zwwang98) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java]: > @zwwang98 allow reviewer to resolve all items as they come back to PR, simply reply "done" to each item as you complete it.

Thank you @nrmancuso , it is really helpful!

Comment (zwwang98) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java (hunk 2 above)
Comment (zwwang98) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java]: done

**Comment by romani** [comment]
GitHub, generate site

**Comment by romani** [comment]
@zwwang98, please do not resolve review items, only reviewer should do this as confirmation that it is done. You should do only comment "done"

**Comment by zwwang98** [comment]
> @zwwang98, please do not resolve review items, only reviewer should do this as confirmation that it is done. You should do only comment "done"

Noted! Will not do that again!

**Comment by romani** [comment]
GitHub, generate site

**Comment by romani** [comment]
Ok to merge.

**Review by romani — APPROVED** [review]

**Pull request merged**